Interop/Runtime/src/jvm/kotlin/kotlinx/cinterop/JvmCallbacks.kt

```diff
@@ -60,9 +59,20 @@ private fun getVariableCType(type: KType): CType<*>? {
     }
 }
 
-private val structTypeCache = ConcurrentHashMap<Class<*>, CType<*>>()
+private class Caches {
+    val structTypeCache = mutableMapOf<Class<*>, CType<*>>()
+    val createdStaticFunctions = mutableMapOf<Class<*>, CPointer<CFunction<*>>>()
+}
+
+private val cachesHolder = ThreadLocal<Caches>()
+private val caches: Caches
+    get() = cachesHolder.get() ?: Caches().also { cachesHolder.set(it) }
 
-private fun getStructCType(structClass: KClass<*>): CType<*> = structTypeCache.computeIfAbsent(structClass.java) {
+fun clearJvmCallbackCaches() {
+    cachesHolder.remove()
```

Comment:
I guess we have a memory/resource leak here: cache entries might require manual dispose.